{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5f2c6ce-edc1-46c8-8eab-9000b2b63149",
   "metadata": {},
   "source": [
    "# VCF_functions.py\n",
    "\n",
    "A framework to process variant call files (VCF) and mutation annotation files (MAF) into pandas dataframes. Able to retrieve mutations for proteins of interest using their Ensembl Transcript identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ffdbc4c8-2e7b-471f-b8ef-7e5dd0775f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import VCF_functions as vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65a8d355-9728-4729-972c-c2956b266784",
   "metadata": {},
   "source": [
    "## Process a VCF file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26e3b68a-b21c-4a61-abb2-eec69f02deed",
   "metadata": {},
   "source": [
    "### Without specifying a list of transcript ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c66a8a03-ccc5-4ed6-b4fe-eadafaa629d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vcf = vcf.read_process_vcf('./data/vcf_data/a/a.vcf', 'missense_variant', 'SNV', return_case_id=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ddf2b952-9435-4d43-9e63-2f3e5cbe2c92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NORMAL</th>\n",
       "      <th>TUMOR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1111</td>\n",
       "      <td>.</td>\n",
       "      <td>X</td>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>{'ACGTNacgtnMINUS': '0', 'ACGTNacgtnPLUS': '0'...</td>\n",
       "      <td>X</td>\n",
       "      <td>Y</td>\n",
       "      <td>Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM   POS ID REF ALT QUAL FILTER  \\\n",
       "0  chr7  1111  .   X   Y    .   PASS   \n",
       "\n",
       "                                                INFO FORMAT NORMAL TUMOR  \n",
       "0  {'ACGTNacgtnMINUS': '0', 'ACGTNacgtnPLUS': '0'...      X      Y     Z  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61385e5-d663-4d20-b4c8-a953dbd8fddd",
   "metadata": {},
   "source": [
    "### With a list of specified transcript ids, the information in the `INFO` column for that transcript is expanded and added to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4d177849-9783-4518-92c1-78b54180dcd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = ['ENST00000318560']\n",
    "df_vcf = vcf.read_process_vcf('./data/vcf_data/a/a.vcf', 'missense_variant', 'SNV', return_case_id=False, \n",
    "                              id_list = ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4f2fad29-41ac-4e5e-bd3f-932e353200cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NORMAL</th>\n",
       "      <th>...</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "      <th>PUBMED</th>\n",
       "      <th>MOTIF_NAME</th>\n",
       "      <th>MOTIF_POS</th>\n",
       "      <th>HIGH_INF_POS</th>\n",
       "      <th>MOTIF_SCORE_CHANGE</th>\n",
       "      <th>ENTREZ</th>\n",
       "      <th>EVIDENCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1111</td>\n",
       "      <td>.</td>\n",
       "      <td>X</td>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>{'ACGTNacgtnMINUS': '0', 'ACGTNacgtnPLUS': '0'...</td>\n",
       "      <td>X</td>\n",
       "      <td>Y</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5243</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM   POS ID REF ALT QUAL FILTER  \\\n",
       "0  chr7  1111  .   X   Y    .   PASS   \n",
       "\n",
       "                                                INFO FORMAT NORMAL  ...  \\\n",
       "0  {'ACGTNacgtnMINUS': '0', 'ACGTNacgtnPLUS': '0'...      X      Y  ...   \n",
       "\n",
       "  CLIN_SIG SOMATIC PHENO PUBMED MOTIF_NAME MOTIF_POS HIGH_INF_POS  \\\n",
       "0                                                                   \n",
       "\n",
       "  MOTIF_SCORE_CHANGE ENTREZ EVIDENCE  \n",
       "0                      5243           \n",
       "\n",
       "[1 rows x 77 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1150169-739e-4ac4-b0ed-442a0d613eb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Consequence</th>\n",
       "      <th>IMPACT</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Feature</th>\n",
       "      <th>BIOTYPE</th>\n",
       "      <th>EXON</th>\n",
       "      <th>INTRON</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1130</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000622132.3:c.2428A&gt;G</td>\n",
       "      <td>ENSP00000478255.1:p.Thr810Ala</td>\n",
       "      <td>2753/4778</td>\n",
       "      <td>2428/3843</td>\n",
       "      <td>1/1130</td>\n",
       "      <td>T/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Consequence    IMPACT SYMBOL             Gene Feature_type  \\\n",
       "0  missense_variant  MODERATE   ABL1  ENSG00000085563   Transcript   \n",
       "\n",
       "           Feature         BIOTYPE    EXON INTRON  \\\n",
       "0  ENST00000318560  protein_coding  1/1130          \n",
       "\n",
       "                         HGVSc                          HGVSp cDNA_position  \\\n",
       "0  ENST00000622132.3:c.2428A>G  ENSP00000478255.1:p.Thr810Ala     2753/4778   \n",
       "\n",
       "  CDS_position Protein_position Amino_acids  \n",
       "0    2428/3843           1/1130         T/A  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf[['Consequence', 'IMPACT', 'SYMBOL', 'Gene',\n",
    "       'Feature_type', 'Feature', 'BIOTYPE', 'EXON', 'INTRON', 'HGVSc',\n",
    "       'HGVSp', 'cDNA_position', 'CDS_position', 'Protein_position',\n",
    "       'Amino_acids']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1cc6b11-6cfc-4786-9381-8263d62a0c68",
   "metadata": {},
   "source": [
    "## Process a MAF file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b994815e-0cb5-4865-ae1c-8db696abecf5",
   "metadata": {},
   "source": [
    "### Without specifying a list of transcript ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d33566a1-814c-4e8e-a122-3556c93118bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_maf = vcf.read_process_maf('./data/maf_data/a.maf', variant='Missense_Mutation', consequence='missense_variant', \n",
    "                              variant_class='SNP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c0a09004-fd9f-4cf8-ad4c-f567439eb944",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>normal_bam_uuid</th>\n",
       "      <th>case_id</th>\n",
       "      <th>GDC_FILTER</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>hotspot</th>\n",
       "      <th>RNA_Support</th>\n",
       "      <th>RNA_depth</th>\n",
       "      <th>RNA_ref_count</th>\n",
       "      <th>RNA_alt_count</th>\n",
       "      <th>callers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATPAF1</td>\n",
       "      <td>64756</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46665282</td>\n",
       "      <td>46665282</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>d4d55edf-e3ee-4b91-8b03-87afdf9bfb5a</td>\n",
       "      <td>bfd49783-1767-469b-9d79-7822301c5efc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COSM426393</td>\n",
       "      <td>N</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>muse;mutect2;varscan2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RORC</td>\n",
       "      <td>6097</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>151814660</td>\n",
       "      <td>151814660</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>d4d55edf-e3ee-4b91-8b03-87afdf9bfb5a</td>\n",
       "      <td>bfd49783-1767-469b-9d79-7822301c5efc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COSM423839;COSM4813058</td>\n",
       "      <td>N</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>muse;mutect2;varscan2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "1      ATPAF1           64756  WUGSC     GRCh38       chr1        46665282   \n",
       "2        RORC            6097  WUGSC     GRCh38       chr1       151814660   \n",
       "\n",
       "   End_Position Strand Variant_Classification Variant_Type  ...  \\\n",
       "1      46665282      +      Missense_Mutation          SNP  ...   \n",
       "2     151814660      +      Missense_Mutation          SNP  ...   \n",
       "\n",
       "                        normal_bam_uuid                               case_id  \\\n",
       "1  d4d55edf-e3ee-4b91-8b03-87afdf9bfb5a  bfd49783-1767-469b-9d79-7822301c5efc   \n",
       "2  d4d55edf-e3ee-4b91-8b03-87afdf9bfb5a  bfd49783-1767-469b-9d79-7822301c5efc   \n",
       "\n",
       "  GDC_FILTER                  COSMIC  hotspot RNA_Support RNA_depth  \\\n",
       "1        NaN              COSM426393        N     Unknown       NaN   \n",
       "2        NaN  COSM423839;COSM4813058        N     Unknown       NaN   \n",
       "\n",
       "   RNA_ref_count  RNA_alt_count                callers  \n",
       "1            NaN            NaN  muse;mutect2;varscan2  \n",
       "2            NaN            NaN  muse;mutect2;varscan2  \n",
       "\n",
       "[2 rows x 140 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_maf.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd24ebd1-3386-48e9-9fad-df9b7e954295",
   "metadata": {},
   "source": [
    "The filtering is done on the `all_effects` column, there may be some variants reported as Silent for the canonical transcript but they are missense for an alternative transcript, those cases are kept"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beafb9b8-5b39-4d0e-ad20-e58bf0b19ae4",
   "metadata": {},
   "source": [
    "### With a list of specified transcript ids, the information in the `all_effects` column for that transcript is expanded and added to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "17dc918d-1001-4bae-bff9-9076d6367396",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = ['ENST00000574428', 'ENST00000318247']\n",
    "df_maf = vcf.read_process_maf('./data/maf_data/a.maf', variant='Missense_Mutation', consequence='missense_variant', \n",
    "                              variant_class='SNP', id_list=ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a672ef4f-25e9-4d3a-b55e-cb08c8a6e48c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>Consequence_all_effects</th>\n",
       "      <th>HGVSp_Short_all_effects</th>\n",
       "      <th>Transcript_ID_all_effects</th>\n",
       "      <th>RefSeq_all_effects</th>\n",
       "      <th>HGVSc_all_effects</th>\n",
       "      <th>Impact_all_effects</th>\n",
       "      <th>Canonical_all_effects</th>\n",
       "      <th>Sift_all_effects</th>\n",
       "      <th>PolyPhen_all_effects</th>\n",
       "      <th>Strand_all_effects</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATPAF1</td>\n",
       "      <td>64756</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46665282</td>\n",
       "      <td>46665282</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.D117N</td>\n",
       "      <td>ENST00000574428</td>\n",
       "      <td></td>\n",
       "      <td>c.349G&gt;A</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td></td>\n",
       "      <td>deleterious(0.05)</td>\n",
       "      <td>possibly_damaging(0.834)</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RORC</td>\n",
       "      <td>6097</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>151814660</td>\n",
       "      <td>151814660</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.E283K</td>\n",
       "      <td>ENST00000318247</td>\n",
       "      <td>NM_005060.4</td>\n",
       "      <td>c.847G&gt;A</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>YES</td>\n",
       "      <td>deleterious(0)</td>\n",
       "      <td>benign(0.036)</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 151 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \\\n",
       "0      ATPAF1           64756  WUGSC     GRCh38       chr1        46665282   \n",
       "1        RORC            6097  WUGSC     GRCh38       chr1       151814660   \n",
       "\n",
       "   End_Position Strand Variant_Classification Variant_Type  ...  \\\n",
       "0      46665282      +      Missense_Mutation          SNP  ...   \n",
       "1     151814660      +      Missense_Mutation          SNP  ...   \n",
       "\n",
       "  Consequence_all_effects HGVSp_Short_all_effects Transcript_ID_all_effects  \\\n",
       "0        missense_variant                 p.D117N           ENST00000574428   \n",
       "1        missense_variant                 p.E283K           ENST00000318247   \n",
       "\n",
       "  RefSeq_all_effects  HGVSc_all_effects Impact_all_effects  \\\n",
       "0                              c.349G>A           MODERATE   \n",
       "1        NM_005060.4           c.847G>A           MODERATE   \n",
       "\n",
       "  Canonical_all_effects   Sift_all_effects      PolyPhen_all_effects  \\\n",
       "0                        deleterious(0.05)  possibly_damaging(0.834)   \n",
       "1                   YES     deleterious(0)             benign(0.036)   \n",
       "\n",
       "   Strand_all_effects  \n",
       "0                  -1  \n",
       "1                  -1  \n",
       "\n",
       "[2 rows x 151 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_maf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eee152a-9237-43d8-9e65-26fc10621b33",
   "metadata": {},
   "source": [
    "When a list of transcript IDs is provided, the data in the `all_effects` column is expanded and added to the dataframe. All the columns coming from the `all_effects` column have '_all_effects' at the end."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ias_env_2]",
   "language": "python",
   "name": "conda-env-ias_env_2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
